### PR TITLE
Use qmk docker image for travis CI builds

### PR DESCRIPTION
**Pull request by zvecr**
## Description

In order to locally reproduce build issues, it is essential to minimise differences. Getting an environment to mimic travis is rather troublesome and involves far too many steps. Using the current docker image for both local builds and CI builds allows contributors to get the same results locally with minimal changes. While it potentially adds the optional docker dependency on the development environment, this still cuts out an entire section of the currently supported platform/compiler matrix.

This PR switches travis to use docker with the following changes:

- Added docker support to the travis configuration
- Removed install of compile dependencies
- Build current qmk_firmware docker image
  - This could pull the one from docker hub if its reliably up to date
  - Alternatively the build could later push this if the image changes
- Added a wrapper script for the test phase
- Force builds to use the make within the qmk docker image when available
  - The current implementation replaces make with a shim but there are other ways to do this if its too hacky

## Types of changes

- [ ] Core
- [ ] Bugfix
- [ ] New Feature
- [x] Enhancement/Optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/Layout/Userspace (addition or update)
- [ ] Documentation

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document. (https://docs.qmk.fm/#/contributing)
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by drashna** [comment]
>In order to locally reproduce build issues, it is essential to minimise differences

Unfortunately, there isn't a good way to do this.  Docker may be a step in the right direction, but .... the versions of AVR-GCC being used is drastically different between OS's.  Additionally, even on the same version, the compiled binaries have different output based on the OS (something I noticed when using AVR-GCC 7.3.0 on Windows and Ubuntu)

That said, using the docker image is a good idea, I think. 

Though, the important thing here, is that the travis script still needs to dump a bunch of the compiled binaries for the firmware.  If it fails to do this, then that is is a problem.

**Comment by zvecr** [comment]
@drashna with the docker volume mount in place, once working, it should not make any differences to the behaviour of the build scripts.

The way i work, i can use whatever platform i want to dev my keymaps, but allows me a quick check before making a PR that i don't get any sudden warnings or failures (even if it is just the "size to big" warning).

**Comment by zvecr** [comment]
So i think this is all ready for review, happy to tackle the `make all` issues as a separate PR.

**Comment by drashna** [comment]
It looks like the build script if failing to use the "default folder" info, and causing issues with the compile command.  That's why you're seeing the "` keymap default and target default`" errors. 

I think that those need to be fixed, before this could be merged.

**Comment by zvecr** [comment]
@drashna currently working through the failures, #4864 any comments and ideas for the fixes would be more than welcome.

File: util/travis_build.sh [hunk 1]
```diff
@@ -1,5 +1,12 @@
 #!/bin/bash
 
+# if docker is installed - call make within the qmk docker image
+if docker -v &> /dev/null; then
+    function make() {
+      docker run --rm -w /qmk_firmware/ -v "$(pwd)":/qmk_firmware qmkfm/qmk_firmware make "$@"
```
Comment (vomindoraan) [inline util/travis_build.sh]:
```suggestion
      docker run --rm -w /qmk_firmware/ -v "$PWD":/qmk_firmware qmkfm/qmk_firmware make "$@"
```
It's safe to use `$PWD` here.

File: util/travis_test.sh [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+# if docker is installed - call make within the qmk docker image
+if docker -v &> /dev/null; then
+    function make() {
+      docker run --rm -w /qmk_firmware/ -v "$(pwd)":/qmk_firmware qmkfm/qmk_firmware make "$@"
```
Comment (vomindoraan) [inline util/travis_test.sh]:
```suggestion
      docker run --rm -w /qmk_firmware/ -v "$PWD":/qmk_firmware qmkfm/qmk_firmware make "$@"
```
Same as above.

File: util/travis_build.sh [hunk 3]
```diff
@@ -1,5 +1,12 @@
 #!/bin/bash
 
+# if docker is installed - call make within the qmk docker image
+if docker -v &> /dev/null; then
```
Comment (vomindoraan) [inline util/travis_build.sh]:
```suggestion
if command -v docker >/dev/null; then
```
For existence checks, I suggest using the `command -v` shell builtin.

File: util/travis_test.sh [hunk 4]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+# if docker is installed - call make within the qmk docker image
+if docker -v &> /dev/null; then
```
Comment (vomindoraan) [inline util/travis_test.sh]:
```suggestion
if command -v docker >/dev/null; then
```
Same as above.

Comment (zvecr) [reply util/travis_test.sh]: Done

File: util/travis_build.sh (hunk 3 above)
Comment (zvecr) [reply util/travis_build.sh]: Done

File: util/travis_build.sh (hunk 1 above)
Comment (zvecr) [reply util/travis_build.sh]: Done

File: util/travis_test.sh (hunk 2 above)
Comment (zvecr) [reply util/travis_test.sh]: Done

**Comment by zvecr** [comment]
Still unable to progress why the build is behaving slightly differently. This is mainly due to the debug lines i add, not printing out the variables.

```diff
diff --git a/Makefile b/Makefile
index 45e47c0..dca0ff9 100644
--- a/Makefile
+++ b/Makefile
@@ -256,6 +256,7 @@ endef
 define PARSE_RULE
     RULE := $1
     COMMANDS :=
+    $(info CurrentRule: $(RULE))
     # If the rule starts with all, then continue the parsing from
     # PARSE_ALL_KEYBOARDS
     ifeq ($$(call COMPARE_AND_REMOVE_FROM_RULE,all),true)
```
When i add that to the root Makefile it does not print the value, just `CurrentRule: `

Calling all makefile gurus, any idea?

**Comment by zvecr** [comment]
Got it.... its the following in the Dockerfile
```dockerfile
ENV KEYBOARD=ergodox_ez
ENV KEYMAP=default
```

Will investigate for fixes either to the dockerfile or to the build scripts.

**Comment by vomindoraan** [comment]
Although I don't understand how exactly that's causing the issue, feel free to remove those env vars from the Dockerfile — `util/docker_build.sh` isn't using them anyway. As a matter of fact, you should probably change the default command to `CMD make all` as well while you're at it. It makes more sense that way, and `util/docker_build.sh` will be unaffected by the change (and your Travis changes don't look like they'll be affected either).

**Comment by zvecr** [comment]
@vomindoraan that was my default plan, however i also see the following options

1. Change ENV to ARG
    - This stops you running the container like `docker run --rm -e KEYMAP=zvecr -e KEYBOARD=lets_split -w /qmk_firmware/ -v "$PWD":/qmk_firmware qmkfm/qmk_firmware` while this lets you use the built in make command, im not sure how useful this actually is
2. Unset the 2 arguments when running container
    - Something like `docker run --rm -e KEYMAP -e KEYBOARD ...`

My vote is for your suggestion, remove from the dockerfile, but this would affect things outside the CI system.

**Comment by zvecr** [comment]
And to clarify why the `ENV KEYMAP=default` causes issues...

The KEYMAP environment variable set in docker sets the KEYMAP variable within make.
The function `define PARSE_KEYBOARD` then goes through pathway `else ifneq ($$(KEYMAP),)` instead of the usual `else` clause.

**Comment by vomindoraan** [comment]
Yikes. I'm sure you had a blast hunting that down...

Anyway, I don't think removing them will affect anything. We don't recommend (nor mention, for that matter) those vars in the docs. They're a leftover from a previous version. I considered removing them when I was updating the Dockerfile and writing `util/docker_build.sh`, but opted against it as the PR diff was already getting big. Since the build script isn't using them, I think it's safe to remove them.

Just to be safe, I will double-check the other QMK projects to make sure none of them are using the vars either.

**Comment by zvecr** [comment]
It definitely brought back some trauma from before i used cmake.

Thanks for having a look @vomindoraan.

**Comment by vomindoraan** [comment]
qmk_compiler uses Git to check out the code and builds it locally, so it doesn't use the `qmkfm/qmk_firmware` image. So we're good here.

The only occurrences of those two vars are in readmes in qmk_firmware, and all but one of them use the old `edasque/qmk_firmware` image, which still has the variables. In case of that [last one](https://github.com/qmk/qmk_firmware/blob/master/keyboards/planck/keymaps/muzfuz/readme.md), the command isn't correct for the current Dockerfile anyway, so I think we're good to go here as well. ¯\\\_(ツ)\_/¯

@muzfuz You might want to correct the Docker command in your Planck readme to:
`util/docker_build.sh muzfuz:planck`
Or, if you really prefer to call Docker yourself:
`docker run --rm -v "$PWD":/qmk_firmware qmkfm/qmk_firmware make muzfuz:planck`
But I recommend using the script since it works on all platforms.

**Comment by zvecr** [comment]
@drashna we have gone green so should be good for review now.

File: Dockerfile [hunk 5]
```diff
@@ -19,11 +19,8 @@ RUN apt-get update && apt-get install --no-install-recommends -y \
...
 
-ENV KEYBOARD=ergodox_ez
-ENV KEYMAP=default
-
 VOLUME /qmk_firmware
 WORKDIR /qmk_firmware
 COPY . .
 
-CMD make $KEYBOARD:$KEYMAP
+CMD make all:default
```
Comment (vomindoraan) [inline Dockerfile]: Good call on the `:default`.

Comment (zvecr) [reply Dockerfile]: I thought best to align it with what was currently happening in travis_build.sh.

**Comment by zvecr** [comment]
Testing this further locally, i am unable to get the tests to compile and run, no matter what i try (container or not). It also begs to question what is happening under travis, as gtest and gmock are **not** installed within the container.

**Comment by zvecr** [comment]
Anything i can provide (logs,etc) to progress this?

**Comment by drashna** [comment]
To make sure, the hex images that this grabs would still be accessible to the travis scripts, correct?

**Comment by zvecr** [comment]
@drashna as the docker container is mounted with the current `$PWD` all the build artefacts are in the same location. So this change should be transparent to that part of the post build process.

**Comment by vomindoraan** [comment]
@zvecr Hey, could you take a look at the conflicts and resolve them?

**Comment by drashna** [comment]
Just a heads up, the repo addition and force install of dfu-util wouldn't be needed on xenial or newer distros.  This was added, since we need v 0.6 or higher of dfu-util to get dfu-suffix. 

But it may need some testing for docker, to make sure dfu-suffix is installed.

**Comment by zvecr** [comment]
@drashna dfu-util is 0.9 on Debian 9 and dfu-suffix is installed so should be good to go (once there is a solid build on travis).

**Comment by yanfali** [comment]
Having issues with ARM builds. Specifically ` util/docker_build.sh connonkeys/instant60:default`. AVR build with keebio/iris and keebio/iris/rev3 worked great

**Comment by yanfali** [comment]
> Having issues with ARM builds. Specifically ` util/docker_build.sh cannonkeys/instant60:default`. AVR build with keebio/iris and keebio/iris/rev3 worked great

OK, PEBCAK. I used connonkeys not cannonkeys

**Comment by drashna** [comment]
Thanks!

**Pull request merged**